zcash_client_sqlite/src/wallet/init/migrations/shardtree_support.rs

```diff
@@ -128,22 +138,42 @@ impl RusqliteMigration for Migration {
                     )
                 })?;
 
+                if block_height % 1000 == 0 {
+                    debug!(height = block_height, "Migrating tree data to shardtree");
+                }
+                trace!(
+                    height = block_height,
+                    size = block_end_tree.size(),
+                    "Storing Sapling commitment tree size"
+                );
                 stmt_update_block_sapling_tree_size
                     .execute(params![block_end_tree.size(), block_height])?;
 
                 if let Some(nonempty_frontier) = block_end_tree.to_frontier().value() {
-                    shard_tree.insert_frontier_nodes(
-                        nonempty_frontier.clone(),
-                        Retention::Checkpoint {
-                            id: BlockHeight::from(block_height),
-                            is_marked: false,
-                        },
-                    )?;
+                    trace!(
+                        height = block_height,
+                        frontier = ?nonempty_frontier,
+                        "Inserting frontier nodes",
+                    );
+                    shard_tree
+                        .insert_frontier_nodes(
+                            nonempty_frontier.clone(),
+                            Retention::Checkpoint {
+                                id: BlockHeight::from(block_height),
+                                is_marked: false,
+                            },
+                        )
+                        .map_err(|e| match e {
+                            ShardTreeError::Query(e) => ShardTreeError::Query(e),
+                            ShardTreeError::Insert(e) => ShardTreeError::Insert(e),
+                            ShardTreeError::Storage(_) => unreachable!(),
```

Comment:
Why is this unreachable? [`ShardTree::insert_frontier_nodes`](https://github.com/zcash/incrementalmerkletree/blob/bae25ad89c0c192bee625252d2d419bd56638e48/shardtree/src/lib.rs#L539) does seem to be able to return `ShardTreeError::Storage`, or at least it maps errors from functions it calls on its `store` into that.

Reply:
As I said above, `CachingShardStore` only ever returns `Infallible` from its `ShardStore` trait impl. So they are by definition unreachable, and due to the type parameterisation any `ShardTreeError::Storage` is therefore also unreachable.